packages/doenetml-worker-rust/doenetml-core/src/state_var_calculations.rs

```diff
@@ -0,0 +1,401 @@
+    // with thousands of levels in the dependency graph, and it wasn't clear what
+    // size WASM stack would be appropriate.
+
+    let original_freshness = components[original_state_var_ptr.component_idx]
+        .borrow()
+        .get_state_variables()[original_state_var_ptr.state_var_idx]
+        .get_freshness();
+
+    // If the current state variable is fresh, there's nothing to do.
+    // If it is unresolved, resolve it
+    match original_freshness {
+        Freshness::Fresh => return (),
+        Freshness::Unresolved => {
+            resolve_state_var(
+                original_state_var_ptr,
+                components,
+                dependencies,
+                dependent_on_state_var,
+                dependent_on_essential,
+                essential_data,
+                should_initialize_essential_data,
+            );
+        }
+        Freshness::Stale | Freshness::Resolved => (),
+    };
+
+    // At this point, the state variable is stale or resolved.
+
+    freshen_stack.push(original_state_var_ptr);
+
+    while let Some(state_var_ptr) = freshen_stack.pop() {
+        let current_freshness = components[state_var_ptr.component_idx]
+            .borrow()
+            .get_state_variables()[state_var_ptr.state_var_idx]
+            .get_freshness();
+
+        // If we have cycles in the graph, it's possible that this
+        // state variable was reached from another path and already freshened.
+        // Skip it in that case.
+        match current_freshness {
+            Freshness::Fresh => continue,
+            Freshness::Unresolved => {
+                panic!("Should not have an unresolved state variable here!")
+            }
+            Freshness::Stale | Freshness::Resolved => (),
+        };
+
+        let dependencies_for_state_var =
+            &dependencies[state_var_ptr.component_idx][state_var_ptr.state_var_idx];
+
+        let mut found_stale_or_resolved_dependency = false;
+
+        for deps in dependencies_for_state_var.iter() {
+            for dep in deps.iter() {
+                match &dep.source {
+                    DependencySource::StateVar {
+                        component_idx,
+                        state_var_idx,
+                    } => {
+                        let new_state_var_ptr = StateVarPointer {
+                            component_idx: *component_idx,
+                            state_var_idx: *state_var_idx,
+                        };
+
+                        let new_current_freshness = dep.value.get_freshness();
+
+                        match new_current_freshness {
+                            // No need to recurse if the state var of the dependency is already fresh
+                            // so don't do anything to just continue the inner loop
+                            Freshness::Fresh => (),
+                            Freshness::Stale | Freshness::Resolved => {
+                                if !found_stale_or_resolved_dependency {
```

Comment:
This is really tracking whether `state_var_ptr` has already been re-added to the stack. Maybe call this variable `state_var_ptr_on_stack`? And set it to false as the first thing you do in the loop.

Reply:
Changed the variable name and added more comments. Hope it is clearer now.